packages/net/tcp_connection.pony

```diff
@@ -955,3 +961,6 @@ actor TCPConnection
       _throttled = false
       _notify.unthrottled(this)
     end
+
+  fun get_fd(): U32 =>
```

Comment:
I still believe that the `fd` should not be exposed in this way - it's not really in line with the capability security model.

As I mentioned in the sync call, I'd prefer to see the "set socket option" and "get socket option" methods be moved here to `TCPConnection`, so that the `fd` can remain hidden.

Reply:
The problem I ran into is that `addressof` can only be used for arguments to an FFI call.  Is there a way to permit a pointer to be passed through the `TCPConnection` object boundary?  I couldn't find a way, but I'm still new at that game, so I may have overlooked something.

Many useful `getsockopt(2)` and `setsockopt(2)` calls involve integers at 4 bytes.  Creating a kludge where those things are copied across `TCPConnection` object boundary is a hack that I know how to do, using a `Reader` and `Writer` to .  For any more complicated ones, such as `IP_ADD_MEMBERSHIP` and `IP_DROP_MEMBERSHIP` at https://github.com/ponylang/ponyc/blob/master/src/libponyrt/lang/socket.c#L1232-L1242 I don't know how to do in pure Pony strictly inside of `TCPConnection` without also resorting to `Reader` and `Writer`.

Reply:
To be clear about what I'm envisioning:
- `OSSocket` goes away
  - `option_size_changed_error` gets moved probably to `OSSockOpt`
  - the rest of the functions in `OSSocket` get moved to be methods of `TCPConnection`, with the `fd` taken from the field instead of passed as an argument.

I don't think the above plan would run into any issues with `addressof`?

Reply:
To be clear, the capability security issue I have with exposing the `fd` is that it is immutable in theory (a `val` reference in the type system) that in practice lets you do mutation on the socket. This is problematic for trying to enforce capability restrictions through the type system.

Reply:
I've cooked up an example at http://playground.ponylang.org/?gist=59fb7dfdf9afbaa7798ad9b41b439735 of the compiler error that I don't know how to work around:

    0.21.3 [release]
    compiled with: llvm 3.9.1 -- cc (Ubuntu 5.4.0-6ubuntu1~16.04.5) 5.4.0 20160609
    Error:
    main.pony:11:7: the addressof operator can only be used for FFI arguments
          addressof optval,
          ^
    Error:
    main.pony:12:7: the addressof operator can only be used for FFI arguments
          addressof optlen)
          ^

Reply:
I might be missing something - I don't understand why you can't just pick up the implementation of `OSSocket`'s `_get_so` and `_set_so` and plop them directly into `TCPConnection` (without the `fd` argument).

If your implementation works in `OSSocket`, I don't see why it can't work when just moved to `TCPConnection`?

Reply:
1. I'd need to cut-and-paste those functions into `udp_socket.pony` also, which wouldn't be the end of the world, but,
2. `_get_so()` and `_set_so()` can only deal with 32 bit values.  Larger ones are pretty common, for example, passing pointers to (larger than 32 bit) C structures, such as the C code needed for `IP_ADD_MEMBERSHIP`, mentioned earlier today.

Reply:
Hrm, passing a lambda across the the `TCPAlike` (playground) or `TCPConnection` real boundary might be doable, but my brain is too fried to try tonight. And that feels like quite a kludge for an API, if it looks like I imagine it might.  {shrug}

Reply:
Okay, the part I was missing was the part where you were still expecting some use cases to require the FFI interface instead of using `OSSocket`.

I took another look at your playground link. I still don't think the `addressof` is necessary to pass to the pony method - the same can be emulated by taking it as a normal argument and returning it in the return value. Or in cases like the one you had with an I32 option, there's no reason to pass one in, so you can just return it.

I noticed that you were assuming `getsockopt` would be a behaviour, but I don't think it should be - it should be a method that requires synchronous access to the `TCPConnection`, which only `TCPConnectionNotify` has. So I changed this in your playground link too.

https://playground.ponylang.org/?gist=59fb7dfdf9afbaa7798ad9b41b439735

Reply:
> I'd need to cut-and-paste those functions into udp_socket.pony

If you like, you could keep the FFI-wrapping implementations in a private primitive, which both `TCPConnection` and `UDPSocket` would have thin wrapper methods for, where they just pass their `_fd` in as the first argument.

> Larger ones are pretty common, for example, passing pointers to (larger than 32 bit) C structures, such as the C code needed for IP_ADD_MEMBERSHIP, mentioned earlier today.

These can be wrapped as well - they just need a wrapper method that accounts for having the right signature.

Reply:
After a chat with @SeanTAllen where neither of us are 100% clear of what you mean by a wrapper method, I'm wondering if pestering you later this week via chat might be possible?

One way I could interpret wrapper methods for right/proper signatures would be one wrapper per data size returned by the kernel.  One example would be Linux's `getsockopt( tcp_work_socket, SOL_IP, TCP_INFO, ...)` where the value stored in arg4's pointer is a `struct tcp_info`, https://github.com/torvalds/linux/blob/2a17178/include/uapi/linux/tcp.h#L168-L227 for its value in `master` branch of Linus's tree today.  Is the kind of wrapper that you're referring handle each of the members of that structure?

Reply:
Maybe we should chat about it during or after the sync call.

However, to illustrate my idea, I created a playground link that shows example method signatures I'd expect to see: https://playground.ponylang.org/?gist=691db79876e8dea099be6f4f509f4c73

Reply:
Hmm, that's about what I thought you were suggesting.  This feels like a case of a 80-20 or 90-10 rule.  Lots of options are a single word, and I'd initially created helpers for that case.

But the remaining cases, there are many dozens hiding in the magic constants that I added to `pony_os_sockopt_option`, and likely more for anyone experimenting with network protocols in a custom kernel/kernel module(*).  Even the `struct tcp_info` case that we've been discussing in the last day could likely change when someone decides to add support for a new RFC and adds even more members to that structure.

(*)  That person probably also would not mind (too much) putting the magic constants into their Pony code rather than discovering them in a platform-independent way via `@pony_os_sockopt_option`.